setup.py

```diff
@@ -1,8 +1,96 @@
 #!/usr/bin/env python
 # -*- coding: utf-8 -*-
 
-from setuptools import setup, find_packages
+import os
+import sys
+from setuptools import setup, find_packages, Extension, Command
+from setuptools.command.build_ext import build_ext
 
+
+# ---------------------------------------------------------------------------- #
+#
+def join_path_list(path_list):
+    path = path_list[0]
+    path_list.remove(path)
+    for _path in path_list:
+        path = os.path.join(path, _path)
+    return path
+
+
+# ---------------------------------------------------------------------------- #
+#
+def TomoPyLibraryFile():
+    """
+    Determines the library output file
+    """
+
+    libname = None
+    if sys.platform.lower().startswith('win'):
+        libname = 'libtomopy.dll'
+    elif sys.platform == 'darwin':
+        libname = 'libtomopy.dylib'
+    else:
+        libname = 'libtomopy.so'
+
+    path = join_path_list([os.getcwd(), "tomopy", "sharedlibs", libname])
+    if os.path.exists(path):
+        return path
+    else:
+        path = join_path_list([os.getcwd(), "tomopy", "sharedlibs", "libtomopy"])
+        for ext in ["so", "dylib", "dll"]:
+            tmp = "{}.{}".format(path, ext)
+            if os.path.exists(tmp):
+                return tmp
+
+    # return best guess
+    return join_path_list([os.getcwd(), "tomopy", "sharedlibs", libname])
+
+
+# ---------------------------------------------------------------------------- #
+#
+class TomoPyExtension(Extension):
+
+    def __init__(self, name, sourcedir=''):
+        Extension.__init__(self, name, sources=[])
+        self.sourcedir = os.path.abspath(sourcedir)
+
+
+# ---------------------------------------------------------------------------- #
+#
+class TomoPyBuild(build_ext, Command):
+
+    #--------------------------------------------------------------------------#
+    # run
+    def run(self):
+        for ext in self.extensions:
+            self.build_extension(ext)
+
+    #--------------------------------------------------------------------------#
+    # build extension
+    def build_extension(self, ext):
+        import build as builder
+        curpath = os.getcwd()
+        os.chdir('config')
+        builder.build_libtomopy()
+        os.chdir(curpath)
+        extdir = os.path.abspath(
+            os.path.dirname(self.get_ext_fullpath(ext.name)))
+        src = TomoPyLibraryFile()
+        src_base = os.path.basename(src)
+        dst = join_path_list([extdir, "tomopy", "sharedlibs", src_base])
+        print("")
+        print('Library: {}'.format(src))
+        print('Destination: {}'.format(dst))
+        print("")
+        try:
+            import shutil
+            shutil.copy2(src, dst)
+        except Exception as e:
+            print("Exception occurred copying '{}' to '{}'... {}".format(src, dst, e))
+
+
+# ---------------------------------------------------------------------------- #
+#
```

Comment:
Minor point: why not use `os.path.join(os.getcwd(), "tomopy", "sharedlibs", libname)` instead of your `join_path_list`?

Larger point:  All this code is added to replace `python build.py` as an initial step.   If that is  actually needed,  maybe consider something like `subprocess.check_call(('python', 'build.py'))`.  That would be a lot less new code.

Reply:
About the larger point, can you explain why this addition is needed instead of just calling `build.py` from inside `setup.py`?

Reply:
@newville w.r.t. minor point -- I'll fix that. Used to more C/C++ so when it didn't accept a list, I just assumed I'd have to do that

Reply:
w.r.t. larger point

- I figured this style was the better option since it gives access to the standard setup.py installation environment
- I can modify it to that scheme if you want but, in the end, I might not matter if y'all want to integrate a CMake build system (I have this done elsewhere)
- CMake will enable a number of improvements over the current `build.py` system, here are a few off the top of my head:  
    - Improved platform portability for compiled code
    - Improve compiler support  (MSVC, Clang, PGI, etc.)
    - More customizable build options without dependence on knowing specific compiler flags
    - Better package support configuration (e.g. enable MPI, CUDA, OpenMP, etc.) under-the-hood for compiled code
    - Easily enable C++ support
    - It will be much easier to generate code coverage for C/C++ code
    - Ability to run clang-tidy on source code

Reply:
@jrmadsen Using CMake instead of Make sounds like it has a lot of benefits, but I believe that switch is independent of the changes that you have made to `setup.py`. You have added a bunch of classes to in order to use function calls intended for the `setuptools.Extension` module, but we don't have a Python extension. We are using `ctypes` to load an external library. Could you please explain benefit of telling `setuptools` that we have an extension when we don't actually have an extension?

Reply:
I am pretty sure a Python extension is just defined as any compiled code with Python wrappers. So I used `setuptools.Extension` because thats how you create a Python compiled extension.

Reply:
@jrmadsen 

> w.r.t. larger point
> I figured this style was the better option since it gives access to the standard setup.py 
> installation environment I can modify it to that scheme if you want but, in the end, I 
> might not matter if y'all want to integrate a CMake build system (I have this done elsewhere)
> CMake will enable a number of improvements over the current build.py system, 

The arguments you make here are advocating for CMake.  I think the concern with the code added to `setup.py` is that it uses way more than 10 lines of code and many levels of indent (that is, complexity) to replace `python build.py` or something like:

```python
    # add this to setup.py?
    from build import buildtomopy
    os.chdir('config')
    build_libtomopy()
    os.chdir('..')
```

That seems completely independent from whether CMake is used or not.

> I am pretty sure a Python extension is just defined as any compiled code with Python 
> wrappers. So I used setuptools.Extension because thats how you create a Python 
> compiled extension.

Well, sort of.  Setuptools seems to expect that an extensions includes python calls in C (or similar) code.  Setuptools will compile the Python+C extensions, often with a `pyd` extension.  That is appropriate for compiled code that really does "<include Python.h>".  When using Cython or ctypes, the definition of extension get blurry.  

It turns out that the compiled code in tomopy does not use Python.h at all and compiles straight C code to a shared library that could be used by (in principle) any language, and is used in tomopy by importing that shared library with `ctypes`.  In that sense,  libtomopy is as much a Python extension as libc is.

So, you can consider libtomopy an extension.  OTOH, expecting that setuptools will automagically compile the code has proven pretty hard, especially on Windows.  If you do want to alter the scripts to build `libtomopy` to build with CMake or to use setuptools.Extension, please make sure the resulting library compiles and works on Windows.

Reply:
## Ok, I hope this clears everything up

You cannot just add these lines to TomoPy. 

```python
    from build import buildtomopy
    os.chdir('config')
    build_libtomopy()
    os.chdir('..')
```

I know it seems like you can, you probably tested it and all the unit tests passed so everything seems fine. Now leave that directory and try to run something with tomopy. _**Everything will fail**_ because that "simple" addition does not result in libtomopy actually getting installed. THAT is why I need to do the copy. I did not just add in a bunch of "excess" code. It is entirely necessary to add those extra layers so that you know exactly where you need to copy libtomopy so that it gets installed. If you don't realize why this is the case it is because when separated into two different commands (build and install) libtomopy exists in `tomopy/sharedlibs` when install gets invoked so `libtomopy` is included when setuptools globs the files within `tomopy` -- which setuptools recognizes as the package folder since it matches the name specified in `setup(name="tomopy", ...)`. When everything is in one step, libtomopy does not exist (because it hasn't been built yet) so it is not included in the install manifest because it has absolutely zero way of know that file will exist eventually. Thus I had to add all this "Extension" code to get the path that I need in order to copy the compiled library into the folder that setuptools installs.

> Well, sort of. Setuptools seems to expect that an extensions includes python calls in C (or similar) code

setuptools makes no such assumption unless you are trying to use setuptools to compile code -- which it absolutely not the case here.

> It turns out that the compiled code in tomopy does not use Python.h at all and compiles straight C code to a shared library that could be used by (in principle) any language, and is used in tomopy by importing that shared library with ctypes. 

You only need `Python.h` if you are making Python interfaces to **_objects_**. Function calls are not objects and tomopy works the way it does without the need for `Python.h` because it only uses fundamental types -- `int`, `float`, etc. -- and numpy arrays which under the hood are just pointers to a starting memory address for a contiguous block of fundamental types (`int*`, `float*`, etc.).

> In that sense, libtomopy is as much a Python extension as libc is.

You can create Python bindings to functions defined in libc and that package would be a Python extension. It is not the library itself that is the extension, it is the combination of the bindings and the library.

> So, you can consider libtomopy an extension. OTOH, expecting that setuptools will automagically compile the code has proven pretty hard, especially on Windows. If you do want to alter the scripts to build libtomopy to build with CMake or to use setuptools.Extension, please make sure the resulting library compiles and works on Windows.

Again, none of my additions to setup.py have anything to do with setuptools compiling anything at all. Those PyCTest and TiMemory packages added to Travis are my packages. Both of these packages are:

- installable via `python setup.py install`
- use setuptools in general but use setuptools to compile absolutely nothing
- entirely use CMake to build the packages
- work with GCC, Clang, Intel, PGI, and MSVC compilers
    - MinGW is probably supported too, I just never bother because it is much better to use MSVC on Windows

In other words, I have done this sort of thing several times before.

Reply:
Thank you for that detailed explanation. It was helpful to understand your design decisions.

However, I believe we already solve the missing shared library problem by (1) using `pip` and (2) [including the shared libraries in `MANIFST.in`](https://blog.ionelmc.ro/2014/06/25/python-packaging-pitfalls/).

I support of my theory, I just now tested the use of `pip install .` to install `tomopy` from source into a clean `conda` environment after adding the following lines to `setup.py`:

```python
import os
from build import build_libtomopy

os.chdir('config')
build_libtomopy()
os.chdir('..')
```

Then, I checked what was installed in my environment's site-packages for `tomopy`, and I saw `libtomopy.dll` where I expected it to be. Finally, I uninstalled `tomopy` using `pip` and the folder I was viewing disappeared.

Is there a reason that you want/need to call `setup.py install` directly instead of using `pip`? [One of the reasons](https://stackoverflow.com/a/15731459) that using `pip` is better is that it makes uninstalling packages easier because it keeps track of which files were added. The alternative is deleting all of the files manually.